### ISSUE TITLE
Issue when using clusterIPs and backend pods restart

### ISSUE THREAD

**Issue by colinmorelli** [linkerd/linkerd2#2468]
### What is the issue?

*tl;dr*: consuming services talking to backend services via cluster IP (not using hostnames) fail first requests after backends restart

--

@grampelberg helped track down an issue today in slack where one of our front end services would fail to call to a backend service on the first call after the backend pods restart (i.e. during a deployment).

After exhausting a number of possibilities, we narrowed it down to the fact that the consuming service was trying to call to a virtual IP instead of using the DNS hostname of the backend service. As a result, linkerd was unable to determine the list of endpoints (since the only address it was aware of was the clusterIP for the service). For reasons that are probably beyond my understanding of linkerd at this time, the service call would fail, and linkerd had no other endpoints to fall back on (because the only "endpoint" it was aware of is the clusterIP)

### How can it be reproduced?

Create a consuming service with linkerd sidecar, a backend service with linkerd, and a clusterIP service in front of the backend.

Point the consuming service at the clusterIP of the backend, without using host headers.

### Logs, error output, etc

Application logs:

```Error: 1 CANCELLED: Received http2 header with status: 500```

The relevant errors in the linkerd2 proxy logs:

```DBUG 10.5.39.103:80 tower_balance updating from discover
DBUG 10.5.39.103:80 tower_balance choosing from 1 replicas
DBUG tokio_reactor loop process - 1 events, 0.000s
DBUG 10.5.39.103:80 linkerd2_proxy::proxy::http::client client request: method=POST uri=http://10.5.39.103:80/serviceEndpoint version=HTTP/2.0 headers={"authorization": "--", "te": "trailers", "content-type": "application/grpc", "user-agent": "grpc-node/1.19.0 grpc-c/7.0.0 (linux; chttp2; gold)", "grpc-accept-encoding": "identity,deflate,gzip", "accept-encoding": "identity,gzip", "l5d-dst-canonical": "10.5.39.103:80"}
DBUG tokio_reactor loop process - 1 events, 0.000s
DBUG proxy={client=out dst=10.5.39.103:80 proto=Http2} linkerd2_proxy::proxy::http::h2 http2 conn error: http2 general error: protocol error: not a result of an error
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.6.4.126:37568} linkerd2_proxy::proxy::http::router service error: an operation was canceled internally before starting
DBUG tokio_reactor loop process - 1 events, 0.000s
DBUG proxy={server=out listen=127.0.0.1:4140 remote=10.6.4.126:37568} h2::codec::framed_write send; frame=Headers { stream_id: StreamId(3), stream_dep: None, flags: HeadersFlag { end_stream: true, end_headers: true, padded: false, priority: false } }
DBUG tokio_reactor loop process - 0 events, 0.000s
```

### Environment

- Kubernetes Version: 1.11.5
- Cluster Environment: GKE
- Host OS: GKE
- Linkerd version: 2.2

### Additional context

I'm not sure there really is much that can be done here, besides possibly improving logging and maybe, _maybe_ detecting when services are trying to talk to clusterIPs directly. This was a user-error on my part for sure, but it was difficult and time consuming to track down for a fairly elementary mistake.

**Comment by olix0r** [comment]
This may or may not be related, but this weekend I observed something similar. Before my server pod passed its readiness check, the cluster IP is available for clients to connect to, but no instances exists to accept the connection.  Once the client's connection times out, subsequent attempts will succeed. It's the desired behavior for the proxy to time out individual connections without impacting the request; so we should probably at least write an integration test for the proxy that exercises the buffering & reconnect logic.

**Comment by olix0r** [comment]
@colinmorelli stable-2.3.2 included a number of changes to how the proxy dispatches requests to endpoints, moving buffering, etc. I'd be curious if you can replicate this behavior on a recent linkerd version?
